### DISCUSSION TITLE
Install RTL 0.13.2 failed

### DISCUSSION THREAD

**Discussion by xbukorg** [Ride-The-Lightning/RTL#1147] [Q&A] [answered]
Hi,

I try to install RTL 0.13.2 but it failed after command 

$ npm install --omit=dev

because used unsupported module see attached logfile.

$ node RTL

###
###

internal/modules/cjs/loader.js:880
  const err = new Error(message);
              ^

Error: Cannot find module 'request'
Require stack:
- C:\Users\Daniel_PC\RTL\node_modules\request-promise\lib\rp.js
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:880:15)
    at Function.Module._load (internal/modules/cjs/loader.js:725:27)
    at Module.require (internal/modules/cjs/loader.js:952:19)
    at require (internal/modules/cjs/helpers.js:88:18)
    at C:\Users\Daniel_PC\RTL\node_modules\request-promise\lib\rp.js:11:16
    at module.exports (C:\Users\Daniel_PC\RTL\node_modules\stealthy-require\lib\
index.js:62:23)
    at Object.<anonymous> (C:\Users\Daniel_PC\RTL\node_modules\request-promise\l
ib\rp.js:10:19)
    at Module._compile (internal/modules/cjs/loader.js:1063:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14)
    at ModuleWrap.<anonymous> (internal/modules/esm/translators.js:199:29)
    at ModuleJob.run (internal/modules/esm/module_job.js:152:23)
    at async Loader.import (internal/modules/esm/loader.js:166:24)
    at async Object.loadESM (internal/process/esm_loader.js:68:5) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    'C:\\Users\\Daniel_PC\\RTL\\node_modules\\request-promise\\lib\\rp.js'
  ]
}

$ npm install request --save
npm WARN deprecated request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142
npm WARN deprecated har-validator@5.1.5: this library is no longer supported
npm WARN deprecated uuid@3.4.0: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.

How can I fix it?

Thanx

[2022-12-10T10_23_46_792Z-debug.log](https://github.com/Ride-The-Lightning/RTL/files/10238137/2022-12-10T10_23_46_792Z-debug.log)

**Comment by ShahanaFarooqui** [comment]
Please use ‘npm install request —legacy-peer-deps’ command instead. We are aware about it and will update the documentation soon.

**Reply by xbukorg** [reply]
It not help:

$ npm install request -legacy-peer-deps

npm WARN deprecated request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142
npm WARN deprecated har-validator@5.1.5: this library is no longer supported
npm WARN deprecated uuid@3.4.0: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.
npm ERR! cb() never called!

npm ERR! This is an error with npm itself. Please report this error at:
npm ERR!     <https://npm.community>

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\Daniel_PC\AppData\Roaming\npm-cache\_logs\2022-12-15T15_45_39_496Z-debug.log

**Comment by ShahanaFarooqui** [comment]
Which NodeJS and npm versions are you on? This error seems to be thrown by npm<v6.

**Reply by xbukorg** [reply]
I used:

node-v14.15.4-x64.msi 

git version 2.38.1.windows.1

$ npm version
{
  rtl: '0.13.2-beta',
  npm: '6.14.10',
  ares: '1.16.1',
  brotli: '1.0.9',
  cldr: '37.0',
  icu: '67.1',
  llhttp: '2.1.3',
  modules: '83',
  napi: '7',
  nghttp2: '1.41.0',
  node: '14.15.4',
  openssl: '1.1.1i',
  tz: '2020a',
  unicode: '13.0',
  uv: '1.40.0',
  v8: '8.4.371.19-node.17',
  zlib: '1.2.11'
}

**Reply by xbukorg** [reply]
I try to install Git 2.39.0 instead of Git 2.38.1 maybe help.

**Comment by ShahanaFarooqui** [comment]
Got it. So can you please delete ‘node_modules’ folder in RTL repo and run ‘npm install —omit=dev —legacy-peer-deeps’ this time?

**Reply by xbukorg** [reply]
I do this steps:
 - install Git 2.39.0
 - delete node_modules folder in RTL
 - npm install -omit=dev -legacy-peer-deeps
 
$ npm install -omit=dev -legacy-peer-deeps
npm ERR! cb() never called!

npm ERR! This is an error with npm itself. Please report this error at:
npm ERR!     <https://npm.community>

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\Daniel_PC\AppData\Roaming\npm-cache\_logs\2022-12-15T17_10_52_176Z-debug.log

**Reply by xbukorg** [reply]
I try to install 18.12.1from https://nodejs.org/en/download/ instead of 14.15.4

**Reply by ShahanaFarooqui** [reply]
Only suggestion at this point, is to upgrade your nodeJS and npm. 

I have not seen this `cb() never called!` error with newer versions. It used to be the issue with npm cache and the solution was to clean and verify the cache. But it is not required anymore with newer versions.

**Reply by xbukorg** [reply]
Witch nodejs 18.12.1 I see

$ npm version
{
  rtl: '0.13.2-beta',
  npm: '8.19.2',
  node: '18.12.1',
  v8: '10.2.154.15-node.12',
  uv: '1.43.0',
  zlib: '1.2.11',
  brotli: '1.0.9',
  ares: '1.18.1',
  modules: '108',
  nghttp2: '1.47.0',
  napi: '8',
  llhttp: '6.0.10',
  openssl: '3.0.7+quic',
  cldr: '41.0',
  icu: '71.1',
  tz: '2022b',
  unicode: '14.0',
  ngtcp2: '0.8.1',
  nghttp3: '0.7.0'
}

**Reply by ShahanaFarooqui** [reply]
Looks like you are on npm `8.19.2` now :).

1: Delete RTL's `node_modules`.
2: Delete RTL's package-lock.json.
3: Run `npm install --omit=dev --legacy-peer-deps`

**Reply by xbukorg** [reply]
Yes it is apart of nodejs 18.12.1

**Reply by xbukorg** [reply]
The result is:

$ npm install -omit=dev -legacy-peer-deeps
npm WARN deprecated csurf@1.11.0: Please use another csrf package
npm WARN deprecated request-promise@4.2.6: request-promise has been deprecated because it extends the now deprecated request package, see https://github.com/request/request/issues/3142
npm WARN deprecated har-validator@5.1.5: this library is no longer supported
npm WARN deprecated sourcemap-codec@1.4.8: Please use @jridgewell/sourcemap-codec instead
npm WARN deprecated uuid@3.4.0: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.
npm WARN deprecated request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142

added 353 packages, and audited 354 packages in 3m

25 packages are looking for funding
  run `npm fund` for details

6 high severity vulnerabilities

To address issues that do not require attention, run:
  npm audit fix

To address all issues (including breaking changes), run:
  npm audit fix --force

Run `npm audit` for details.

Can I start  npm audit fix?

**Comment by ShahanaFarooqui** [comment, marked as answer]
Audit will not fix them, these are known and already reported vulnerabilities. We are still waiting for the library to update them.

**Reply by xbukorg** [reply]
Seems to be fixed :)

$ node RTL
[12/15/2022, 8:08:21 PM] INFO: CORS => Setting up CORS...

[12/15/2022, 8:08:21 PM] INFO: CORS => CORS Set.

[12/15/2022, 8:08:21 PM] INFO: CSRF => Setting up CSRF...

[12/15/2022, 8:08:21 PM] INFO: CSRF => CSRF Set.

[12/15/2022, 8:08:21 PM] INFO: App => Setting up Application Routes...

[12/15/2022, 8:08:21 PM] INFO: App => Application Routes Set.

[12/15/2022, 8:08:21 PM] INFO: WebSocketServer => Connecting Websocket Server...

[12/15/2022, 8:08:21 PM] INFO: WebSocketServer => Websocket Server Connected.

[12/15/2022, 8:08:21 PM] INFO: RTL => Server is up and running, please open the UI at http://localho
st:3000 or your proxy configured url.

**Reply by ShahanaFarooqui** [reply]
Vulnerabilities are not fixed yet but RTL can be functional with them. We will fix them as soon as the library releases their new version.

**Reply by 3nprob** [reply]
@ShahanaFarooqui Upstream update won't happen but there is a maintained fork: https://github.com/Ride-The-Lightning/RTL/pull/1512
